src/openms/source/ANALYSIS/OPENSWATH/MRMFeatureFilter.cpp

```diff
@@ -70,236 +55,176 @@ namespace OpenMS
+        // iterate through feature/sub-feature QCs/filters
+        for (const MRMFeatureQC::ComponentQCs& c_qc : filter_criteria.component_qcs)
         {
-          if (filter_criteria.component_qcs[c_qc_it].component_name == component_name)
+          if (c_qc.component_name == component_name)
           {
-            // RT check
-            double rt = features[feature_it].getSubordinates()[sub_it].getRT(); //check!
-            // std::cout << "RT" << std::endl; //debugging
-            if (!checkRange(rt,
-              filter_criteria.component_qcs[c_qc_it].retention_time_l,
-              filter_criteria.component_qcs[c_qc_it].retention_time_u))
+            if (!isWithinRange((double)subordinate.getRT(), c_qc.retention_time_l, c_qc.retention_time_u))
             {
-              c_qc_pass = false;
+              c_qc_pass[C_Metrics::RT] = false;
               c_qc_fail_message_vec.push_back("retention_time");
             }
-
-            // intensity check
-            // std::cout << "Intensity" << std::endl; //debugging
-            double intensity = features[feature_it].getSubordinates()[sub_it].getIntensity();
-            if (!checkRange(intensity,
-              filter_criteria.component_qcs[c_qc_it].intensity_l,
-              filter_criteria.component_qcs[c_qc_it].intensity_u))
+            if (!isWithinRange((double)subordinate.getIntensity(), c_qc.intensity_l, c_qc.intensity_u))
             {
-              c_qc_pass = false;
+              c_qc_pass[C_Metrics::INTENSITY] = false;
               c_qc_fail_message_vec.push_back("intensity");
             }
-
-            // overall quality check getQuality
-            double quality = features[feature_it].getSubordinates()[sub_it].getOverallQuality();
-            // std::cout << "Quality" << std::endl; //debugging
-            if (!checkRange(quality,
-              filter_criteria.component_qcs[c_qc_it].overall_quality_l,
-              filter_criteria.component_qcs[c_qc_it].overall_quality_u))
+            if (!isWithinRange((double)subordinate.getOverallQuality(), c_qc.overall_quality_l, c_qc.overall_quality_u))
             {
-              c_qc_pass = false;
+              c_qc_pass[C_Metrics::QUALITY] = false;
               c_qc_fail_message_vec.push_back("overall_quality");
             }
-
-            // metaValue checks
-            for (auto const& kv : filter_criteria.component_qcs[c_qc_it].meta_value_qc)
+            for (const std::pair<String, std::pair<double,double>>& kv : c_qc.meta_value_qc)
             {
-              // std::cout << "MetaData" << std::endl; //debugging
-              if (!checkMetaValue(features[feature_it].getSubordinates()[sub_it], kv.first, kv.second.first, kv.second.second))
+              if (!checkMetaValue(subordinate, kv.first, kv.second.first, kv.second.second))
               {
-                c_qc_pass = false;
+                c_qc_pass[C_Metrics::METAVALUE] = false;
```

Comment:
If there are more than 1 metaValue, then the count for the number of passing/failing metavalue checks is lost.

Reply:
I will need to change this. I treated the metavalues as a single test.